racket/collects/racket/private/dict.rkt

```diff
@@ -251,16 +252,28 @@
     (proc (car x) (cdr x))))
 
 (define (assoc-keys d)
-  (for/list ([x (in-list d)])
-    (unless (pair? x)
-      (raise-argument-error 'dict-keys "dict?" d))
-    (car x)))
+  (remove-duplicates
+    (for/list ([x (in-list d)])
+      (unless (pair? x)
+        (raise-argument-error 'dict-keys "dict?" d))
+      (car x))))
+
+(define (assoc-map-unique f d)
+  (let loop ([xd d]
+             [keys (assoc-keys d)])
```

Comment:
In line with @bennn's comment, I think it might turn out to be expensive to implement `assoc-map-unique` in terms of `assoc-keys` and `remove-duplicates`: that seems like a lot of extra `equal?` tests and extra intermediate lists. I'm still thinking about it, but I can imagine an `assoc-fold-unique` that could be used to implement `assoc-count`, `assoc-keys`, `assoc-values`, and `assoc->list`. I think the key insight `remove-duplicates` is to use `member` for lists of less than length 40 and a hash-table otherwise, whereas most of the complexity is from needing to handle general cases:
https://github.com/racket/racket/blob/f8bc4e8fa1d37d90a66e0e975dcbc244cd3c9796/racket/collects/racket/list.rkt#L403-L406

Alternatively, we presumably expect the common case to be lists with no duplicate keys, so maybe it would pay off to avoid allocation in those cases.

That said, I think making these functions correct is good and valuable, even if more could potentially be done to make them efficient.

Reply:
You're right. I've implemented `assoc-fold-unique` which uses an immutable `set` to keep track of the keys seen. The set could be made mutable, but it's unlikely that this would make a significant difference for small alists.

I've re-implemented the `assoc-count`, `assoc-keys`, `assoc-values` and `assoc->list` functions in terms of `assoc-fold-unique`.

Note that this has `foldl` semantics, and will build the resulting lists in reverse, but that should be OK since `dict`s don't guarantee the order anyway.